microbench-ci: demo [dnm]

herkolategan opened the pull request:
Microbenchmark CI demo.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140356)
<!-- Reviewable:end -->

herkolategan committed:
roachprod-microbench: expose parser

Previously, `ExtractBenchmarkResults` was inaccessible by other packages. This
change is a small refactor to allow other packages to use the extract benchmark
results functionality.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: comparison utility

The microbench-ci utility is designed to be used by CI, specifically
microbenchmark related GitHub workflows. It provides functionality for running
microbenchmarks, comparing results, and posting comments to a GitHub PR to
display the results to the author of the PR.

An opinionated selection of benchmarks are specified in `pull-request-suite.yml`
that are stable, in terms of variance, and covers a good surface area. This is
the suite that CI will run for every PR. Care should be taken around the measure
count and iterations to ensure that the running time per group remains under ~20
minutes.

In the event that any of the benchmarks regress by more than the threshold,
profiles will be collected and included in the results.

The benchmarks and profiles run in an interleaved fashion. Any failures will
immediately stop execution and report the errors. Regression does not count as a
failure, and will be displayed in the summary and results on the PR.

Epic: None
Release note: None

herkolategan committed:
dev: add microbench-ci to build

Add build shorthand for `microbench-ci` utility.

Epic: None
Release note: None

herkolategan committed:
github: microbenchmark CI workflow

This change adds GitHub workflows, actions and scripts for running the
microbenchmarks pull request suite specified in the microbench-ci package config
`pull-request-suite.yml`. It will run on every PR and post results to the PR in
the form of a comment (also visible on the job summary).

The jobs first build the head and merge base binaries in parallel, as well as
the `microbench-ci` utility. The next phase runs the microbenchmarks using the
`microbench-ci` utility. The run phases uses groups to parallelize running
different benchmarks and reduce overall CI completion time.

Google cloud storage is used as a binary cache throughout the jobs, and has been
configured to limit the TTL of the cached binaries. The binaries are available
as part of the results in order to reproduce a regression on a GCE worker if
necessary.

Finally, the results are compared and published to GCS. An additional job is
added to post the results. This job is separate and uses the base version of
`microbench-ci` to ensure only this sections has access to the elevated GitHub
token privileges for posting on PRs.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: demo

Disables other CI jobs.
Skips the post.
Sets the workflow to `pull_request` so it runs.

herkolategan committed:
microbench-ci: end to end datadriven test

Create a test that triggers both the run and compare functionalities in
`microbench-ci`. The test mocks benchmark binary outputs, verifying that the
summary templates render correctly and validating the compare logic. Any updates
to the template will require rewriting the testdata.

Epic: None
Release note: None

herkolategan committed:
github: initially disable microbench-ci posting

Initially, the post part of this job will be disabled until the microbenchmark
results produced by the work done in this PR is deemed stable. The job summary
is still available for inspection on the job, but will not yet post a more
visible result to the PR in the form of a comment.

Results on GCS will be inspected over a couple of weeks before enabling the post
job.

Epic: None
Release note: None

herkolategan committed:
microbench-ci: adjust thresholds

Set slightly higher thresholds on
`BenchmarkSysbench/KV/1node_local/oltp_write_only` as this benchmark tends to
have higher variance than read only, and it should be reflected in the config.

Epic: None
Release note: None
